cwe_checker_rs/src/checkers/cwe_676.rs

```diff
@@ -0,0 +1,108 @@
+/// The CWE 676 checks for potentially dangerous function calls, such as strcpy, memcpy, scanf etc.
+
+use crate::{
+    intermediate_representation::{ Program, Project, Sub, Term, Tid, Jmp, ExternSymbol }, 
+    utils::log::{ CweWarning, LogMessage }};
+use serde::{ Serialize, Deserialize };
+
+const VERSION: &str = "0.1";
+
+pub static CWE_MODULE: crate::CweModule = crate::CweModule {
+    name: "CWE676",
+    version: VERSION,
+    run: check_cwe,
+};
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq, Hash, Clone)]
+pub struct Config {
+    symbols: Vec<String>,
+}
+
+
+/// For each block of a subroutine check if the jump is a direct call and if the target TID of that call is equal to a dangerous symbol TID
+pub fn get_call_to_target<'a>(caller: &'a Term<Sub>, target: &'a ExternSymbol) -> Vec<Option<(&'a str, &'a Tid, &'a str)>> {
+    let mut calls: Vec<Option<(&'a str, &'a Tid, &'a str)>> = Vec::new();
+    for blk in caller.term.blocks.iter() {
+        for jmp in blk.term.jmps.iter() {
+            match &jmp.term {
+                Jmp::Call{ target: dst, .. } => {
+                    if *dst == target.tid {
+                        calls.push(Some((caller.term.name.as_str(), &blk.tid, target.name.as_str())));
+                    }
+                }
+                _ => ()
+            }
+        }
+    }
+
+    calls
+}
+
+
+/// For each subroutine and each found dangerous symbol, check for calls to the corresponding symbol
+pub fn get_calls_to_symbols<'a>(subfunctions: &'a Vec<Term<Sub>>, dangerous_symbols: &'a Vec<&'a ExternSymbol>) -> Vec<Option<(&'a str, &'a Tid, &'a str)>> {
+    let mut calls: Vec<Option<(&str, &Tid, &str)>> = Vec::new();
+    for sub in subfunctions.iter() {
+        for symbol in dangerous_symbols.iter() {
+            calls.append(&mut get_call_to_target(sub, symbol))
+        }
+    }
```

Comment:
If you put the dangerous symbols into an appropiate HashMap or HashSet, we would not need the inner loop, which in turn would speed up the theoretical runtime of the algorithm. I don't know why we did not do this on the Ocaml side...